### PR TITLE
Remove defaultProps support (except for classes) 

### PR DESCRIPTION
**Pull request by acdlite**
This removes defaultProps support for all component types except for classes. We've chosen to continue supporting defaultProps for classes because lots of older code relies on it, and unlike function components, (which can use default params), there's no straightforward alternative.

By implication, it also removes support for setting defaultProps on `React.lazy` wrapper. So this will not work:

```js
const MyClassComponent = React.lazy(() => import('./MyClassComponent'));
// MyClassComponent is not actually a class; it's a lazy wrapper. So
// defaultProps does not work.
MyClassComponent.defaultProps = { foo: 'bar' };
```

However, if you set the default props on the class itself, then it's fine.

For classes, this change also moves where defaultProps are resolved. Previously, defaultProps were resolved by the JSX runtime. This change is only observable if you introspect a JSX element, which is relatively rare but does happen.

In other words, previously `<ClassWithDefaultProp />.props.aDefaultProp` would resolve to the default prop value, but now it does not.

### CODE REVIEW

File: packages/react/src/ReactLazy.js [hunk 1]
```diff
@@ -134,53 +135,34 @@ export function lazy<T>(
...
-          );
-          defaultProps = newDefaultProps;
-          // Match production behavior more closely:
-          // $FlowFixMe[prop-missing]
-          Object.defineProperty(lazyType, 'defaultProps', {
-            enumerable: true,
-          });
-        },
-      },
-      propTypes: {
```
Comment (acdlite) [inline packages/react/src/ReactLazy.js:163]: Unrelated, but I removed this warning entirely since `propTypes` support has already been removed from the codebase.

File: packages/react/src/jsx/ReactJSXElement.js [hunk 2]
```diff
@@ -340,12 +344,14 @@ export function jsxProd(type, config, maybeKey) {
...
   }
 
-  // Resolve default props
-  if (type && type.defaultProps) {
-    const defaultProps = type.defaultProps;
-    for (propName in defaultProps) {
-      if (props[propName] === undefined) {
-        props[propName] = defaultProps[propName];
+  if (!disableDefaultPropsExceptForClasses) {
+    // Resolve default props
```
Comment (sebmarkbage) [inline packages/react/src/jsx/ReactJSXElement.js:348]: Shouldn't this be removed even when the flag is off? Because the resolution moves regardless?

**Review by sebmarkbage — APPROVED** [review]

File: packages/react/src/jsx/ReactJSXElement.js (hunk 2 above)
Comment (acdlite) [reply packages/react/src/jsx/ReactJSXElement.js:348]: The way I implemented the flag is that when it's off, it maintains the old behavior: props resolved during JSX. I did it that way to aid the roll out at Meta and reduce the risk the whole change gets reverted. Since there could be code that reads `element.props` and expects the `defaultProps` to already be there.

Comment (acdlite) [reply packages/react/src/jsx/ReactJSXElement.js:348]: So really the flag handles both moving the resolution and removing it for non-classes, I just didn't think it was worth creating two separate flags.
